/*
 * Copyrighted (Kord Extensions, 2024). Licensed under the EUPL-1.2
 * with the specific provision (EUPL articles 14 & 15) that the
 * applicable law is the (Republic of) Irish law and the Jurisdiction
 * Dublin.
 * Any redistribution must include the specific provision above.
 */

package dev.kordex.modules.web.core.backend.server

import dev.kordex.modules.web.core.backend.config.WebServerConfig
import dev.kordex.modules.web.core.backend.server.routes.api
import io.ktor.server.application.*
import io.ktor.server.auth.*
import io.ktor.server.http.content.*
import io.ktor.server.resources.Resources
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.websocket.*

public fun WebServer.configureRouting(app: Application, config: WebServerConfig): ConfiguredRoutes {
	lateinit var extensionApiBaseRoute: Route
	lateinit var pageApiBaseRoute: Route

	app.install(Resources)

	app.routing {
		// TODO: API Routing
		// TODO: Static files

		if (config.devMode) {
			get("/") {
				call.respondRedirect("http://localhost:5173")
			}
		} else {
			if (config.hostname != null && config.siteTitle != null) {
				singlePageApplication {
					useResources = true
					filesPath = "dev/kordex/modules/web/core/frontend"
				}
			}
		}

		if (config.hostname != null) {
			authenticate("oauth-discord") {
				get("/auth") {
					// Redirect is apparently automatic
				}

				get("/auth/callback") {
					val principal: OAuthAccessTokenResponse.OAuth2? = call.principal()

					// TODO: Frontend work, figure out the client-side, handle Discord API stuff, etc

					principal?.let { p ->
						p.state?.let { state ->
							call.respondRedirect(
								"/#auth/callback?state=$state&token=${p.accessToken}"
							)

							return@get
						}
					}

					call.respondRedirect("/#auth/failed")
				}
			}
		}

		extensionApiBaseRoute = route("/api/e") {}
		pageApiBaseRoute = route("/api/p") {}

		route("/ws/e/{path...}") {
			webSocket {
				registries.websockets.handle(this)
			}
		}

		// Bundled routes defined elsewhere

		api(config)
	}

	return ConfiguredRoutes(
		extensionApiBaseRoute,
		pageApiBaseRoute
	)
}

public data class ConfiguredRoutes(
	val extensionApiBase: Route,
	val pageApiBase: Route
)
